pkg/controllers/selection/preferences.go

```diff
@@ -60,22 +67,25 @@ func (p *Preferences) Relax(ctx context.Context, pod *v1.Pod) {
 		p.cache.SetDefault(string(pod.UID), cachedSpec)
 		return
 	}
+
 	// Attempt to relax the pod and update the cache
 	cachedSpec := spec.(v1.PodSpec)
 	pod.Spec.Affinity = cachedSpec.Affinity
 	pod.Spec.Tolerations = cachedSpec.Tolerations
-	if relaxed := p.relax(ctx, pod); relaxed {
+	if relaxed := p.relax(ctx, pod, options); relaxed {
 		p.cache.SetDefault(string(pod.UID), pod.Spec)
 	}
 }
 
-func (p *Preferences) relax(ctx context.Context, pod *v1.Pod) bool {
+func (p *Preferences) relax(ctx context.Context, pod *v1.Pod, options RelaxOptions) bool {
 	for _, relaxFunc := range []func(*v1.Pod) *string{
 		func(pod *v1.Pod) *string { return p.removePreferredPodAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removePreferredPodAntiAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removePreferredNodeAffinityTerm(pod) },
 		func(pod *v1.Pod) *string { return p.removeRequiredNodeAffinityTerm(pod) },
-		func(pod *v1.Pod) *string { return p.toleratePreferNoScheduleTaints(pod) },
+		func(pod *v1.Pod) *string {
+			return p.toleratePreferNoScheduleTaints(pod, options)
+		},
```

Comment:
Do we want to revisit the order in which we relax preferences?  I know when I first pushed this change, the order didn't seem to matter, but maybe it would it make sense to tolerate PreferNoSchedule taints somewhere before relaxing "Required" node affinity terms.

Reply:
It probably makes sense, the prefer no schedule should likely be first as it will be what causes the provisioner to not be acceptable.  The required node affinity terms are removed one at a time, until only one is left, the last is never removed.  Still not perfect, but that's the implementation for now.